### PR TITLE
added encodeArguements to convert string address to ethereum address

### PR DESCRIPTION
**Pull request by koseoyoung**
When we want to deploy/execute a smart contract with the address/[]address argument, it has to convert string address to ethereum address. 

In deploy, execute, readonexecute, I added encodeArgumen() to convert string to ethAddress 
And added test files including address[] argument

### CODE REVIEW

File: iotex/callers.go [hunk 1]
```diff
@@ -382,3 +408,49 @@ func (c *getReceiptCaller) Call(ctx context.Context, opts ...grpc.CallOption) (*
...
+		return nil, errcodes.New("the number of arguments is not correct", errcodes.InvalidParam)
+	}
+	newArgs := make([]interface{}, len(args))
+	for index, input := range method.Inputs {
+		if input.Type.String() == "address" {
+			s, ok := args[index].(string)
+			if !ok {
+				return nil, errcodes.New("fail to convert from interface to string", errcodes.InvalidParam)
+			}
+			ethAddress, err := address.FromString(s)
```
Comment (CoderZhi) [inline iotex/callers.go]: this is actually an ioAddress

File: iotex/callers.go [hunk 2]
```diff
@@ -382,3 +408,49 @@ func (c *getReceiptCaller) Call(ctx context.Context, opts ...grpc.CallOption) (*
...
+			if err != nil {
+				return nil, errcodes.New("fail to convert string to address(iotex)", errcodes.InvalidParam)
+			}
+			newArgs[index] = common.HexToAddress(hex.EncodeToString(ethAddress.Bytes()))
+
+		} else if input.Type.String() == "address[]" {
+			s := reflect.ValueOf(args[index])
+			if s.Kind() != reflect.Slice && s.Kind() != reflect.Array {
+				panic("InterfaceSlice() given a non-slice and non-array type")
```
Comment (CoderZhi) [inline iotex/callers.go]: panic or return error?

File: iotex/callers.go [hunk 3]
```diff
@@ -382,3 +408,49 @@ func (c *getReceiptCaller) Call(ctx context.Context, opts ...grpc.CallOption) (*
...
+			}
+			ret := make([]interface{}, s.Len())
```
Comment (CoderZhi) [inline iotex/callers.go]: do we need this variable?

File: iotex/callers.go [hunk 4]
```diff
@@ -382,3 +408,49 @@ func (c *getReceiptCaller) Call(ctx context.Context, opts ...grpc.CallOption) (*
...
+			for i := 0; i < s.Len(); i++ {
+				ret[i] = s.Index(i).Interface()
+			}
+			newArr := make([]common.Address, s.Len())
+			for j, elem := range ret {
+				str, ok := elem.(string)
+				if !ok {
+					return nil, errcodes.New("fail to convert from interface to string in array", errcodes.InvalidParam)
+				}
+				ethAddress, err := address.FromString(str)
```
Comment (CoderZhi) [inline iotex/callers.go]: this is an ioAddress

File: iotex/callers.go [hunk 5]
```diff
@@ -382,3 +408,49 @@ func (c *getReceiptCaller) Call(ctx context.Context, opts ...grpc.CallOption) (*
...
 }
+
+func encodeArgument(method abi.Method, args []interface{}) ([]interface{}, error) {
+	if len(method.Inputs) != len(args) {
+		return nil, errcodes.New("the number of arguments is not correct", errcodes.InvalidParam)
+	}
+	newArgs := make([]interface{}, len(args))
+	for index, input := range method.Inputs {
+		if input.Type.String() == "address" {
+			s, ok := args[index].(string)
```
Comment (CoderZhi) [inline iotex/callers.go]: string or address? For Ethereum, the type is common.address, correspondingly, we may prefer address, or we can support both.

**Comment by koseoyoung** [comment]
I refactored the code, and made it available to handle both string/address to be an argument

File: iotex/callers.go [hunk 6]
```diff
@@ -382,3 +408,55 @@ func (c *getReceiptCaller) Call(ctx context.Context, opts ...grpc.CallOption) (*
...
+	if len(method.Inputs) != len(args) {
+		return nil, errcodes.New("the number of arguments is not correct", errcodes.InvalidParam)
+	}
+	newArgs := make([]interface{}, len(args))
+	for index, input := range method.Inputs {
+		if input.Type.String() == "address" {
+			str, ok := args[index].(string)
+			if !ok {
+				addr, check := args[index].(address.Address)
+				if !check {
```
Comment (yutongp) [inline iotex/callers.go]: I think we should support both `address.Address` and `common.Address`, if user is already passing `common.Address` it works too

File: iotex/callers.go [hunk 7]
```diff
@@ -382,3 +408,55 @@ func (c *getReceiptCaller) Call(ctx context.Context, opts ...grpc.CallOption) (*
...
+			newArgs[index] = common.HexToAddress(hex.EncodeToString(ioAddress.Bytes()))
+
+		} else if input.Type.String() == "address[]" {
+			s := reflect.ValueOf(args[index])
+			if s.Kind() != reflect.Slice && s.Kind() != reflect.Array {
+				return nil, errcodes.New("fail because the type is non-slice, non-array", errcodes.InvalidParam)
+			}
+			newArr := make([]common.Address, s.Len())
+			for j := 0; j < s.Len(); j++ {
+				str, ok := s.Index(j).Interface().(string)
```
Comment (yutongp) [inline iotex/callers.go]: these are dup logic 419-432, please refactor

File: iotex/callers.go [hunk 8]
```diff
@@ -382,3 +408,55 @@ func (c *getReceiptCaller) Call(ctx context.Context, opts ...grpc.CallOption) (*
...
 }
+
+func encodeArgument(method abi.Method, args []interface{}) ([]interface{}, error) {
+	if len(method.Inputs) != len(args) {
+		return nil, errcodes.New("the number of arguments is not correct", errcodes.InvalidParam)
+	}
+	newArgs := make([]interface{}, len(args))
+	for index, input := range method.Inputs {
+		if input.Type.String() == "address" {
+			str, ok := args[index].(string)
```
Comment (yutongp) [inline iotex/callers.go]: can we use a switch type assertion here:https://tour.golang.org/methods/16 ?

Comment (koseoyoung) [reply iotex/callers.go]: ok!! thanks !!

**Comment by koseoyoung** [comment]
I refactored the code and made it handle ethereum address too

File: iotex/callers.go [hunk 9]
```diff
@@ -382,3 +408,53 @@ func (c *getReceiptCaller) Call(ctx context.Context, opts ...grpc.CallOption) (*
...
 	return c.api.GetReceiptByAction(ctx, &iotexapi.GetReceiptByActionRequest{ActionHash: h}, opts...)
 }
+
+func encodeArgument(method abi.Method, args []interface{}) ([]interface{}, error) {
+	if len(method.Inputs) != len(args) {
+		return nil, errcodes.New("the number of arguments is not correct", errcodes.InvalidParam)
+	}
+	newArgs := make([]interface{}, len(args))
+	for index, input := range method.Inputs {
+		if input.Type.String() == "address" {
```
Comment (yutongp) [inline iotex/callers.go]: nit: switch input.Type.String()

Comment (koseoyoung) [reply iotex/callers.go]: I fixed it thanks !

**Pull request merged**